### PR TITLE
apps sc: add script to restore Harbor from backup

### PR DESCRIPTION
**Pull request by viktor-f**
**What this PR does / why we need it**: This adds the restore script for harbor. I debated putting some documentation here but I though it might be better to just have it on compliantkubernetes.io (PR with that is coming later)

**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*: part of elastisys/compliantkubernetes#77

**Public facing documentation PR** *(if applicable)*

Coming later

**Special notes for reviewer**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

**Comment by OlleLarsson** [comment]
I don't fully understand this script.
Where is this supposed to be run, from localhost or in k8s?

**Comment by viktor-f** [comment]
> I don't fully understand this script -where is this supposed to be run, from localhost or in k8s?

From local host (you can see the old instructions here: https://github.com/elastisys/ck8s-ops/tree/main/backup/harbor). Honestly I think we should rework this script, but that will be done in a later issue.

**Comment by OlleLarsson** [comment]
> > I don't fully understand this script -where is this supposed to be run, from localhost or in k8s?
> 
> From local host (you can see the old instructions here: https://github.com/elastisys/ck8s-ops/tree/main/backup/harbor). Honestly I think we should rework this script, but that will be done in a later issue.

I see.

**Review by OlleLarsson — APPROVED** [review]

File: scripts/restore/restore-harbor.sh [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+extract_backup(){
+    echo "Extracting backups">&2
+    tar xvf harbor.tgz
+}
+
+wait_for_db_ready() {
+    echo "Waiting for DB to be ready" >&2
+    TIMEOUT=12
+    while [ $TIMEOUT -gt 0 ]; do
+        if pg_isready -h $HOSTNAME | grep "accepting connections"; then
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:35]: Is this a dependency for this script (`pg_isready`)? If so, we should add something about it. I'm fine by just mentioning it on top of the file.

File: scripts/restore/restore-harbor.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+s3_download() {
+    : "${S3_BUCKET:?Missing S3_BUCKET}"
+    : "${S3_REGION_ENDPOINT:?Missing S3_REGION_ENDPOINT}"
+    if [[ -n "$1" ]]; then
+        backup_key=$1
+    else
+        backup_key=$(aws s3 ls "${S3_BUCKET}/backups" \
+          --recursive \
+          --endpoint-url="${S3_REGION_ENDPOINT}" \
+          | sort | tail -n 1 | awk '{print $4}')
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:20]: Should we add a prompt here verifying that this is the backup you want to restore from?

File: scripts/restore/restore-harbor.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  echo "Cleaning up local files">&2
+  rm harbor.tgz
+  rm ${backup_dir}/registry.back
+  rm ${backup_dir}/postgres.back
+  rm ${backup_dir}/notarysigner.back
+  rm ${backup_dir}/notaryserver.back
+  rmdir ${backup_dir}
+}
+
+s3_download "$1"
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:79]: Thinking about it, I feel like we should have a prompt before here as well. We are in fact wiping the whole DB, so a warning should make the user at least a bit more aware if not.

File: scripts/restore/restore-harbor.sh [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    done
+    if [ $TIMEOUT -eq 0 ]; then
+        echo "Harbor DB cannot reach within one minute."
+        exit 1
+    fi
+}
+
+clean_database_data() {
+  echo "Droping existing databases">&2
+  psql -U postgres -d template1 -h $HOSTNAME -c "drop database registry;"
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:49]: Same with this, should at least be mentioned in the comments

File: scripts/restore/restore-harbor.sh [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
+# Check https://compliantkubernetes.io/operator-manual/disaster-recovery/ for instructions
+# Restores latest backup from S3_BUCKET. $1 can be used to specify backup
+# To get a list of available backups use:
+# aws s3 ls ${S3_BUCKET}"/backups" --recursive --endpoint-url=$S3_REGION_ENDPOINT
+set -e
+
+HOSTNAME=localhost
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:8]: Should you not be able to set this?

File: scripts/restore/restore-harbor.sh [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+          --endpoint-url="${S3_REGION_ENDPOINT}" \
+          | sort | tail -n 1 | awk '{print $4}')
+    fi
+    echo "Downloading backup from s3 bucket ${backup_key}" >&2
+    aws s3 cp "s3://${S3_BUCKET}/${backup_key}" harbor.tgz --endpoint-url="${S3_REGION_ENDPOINT}"
+}
+
+extract_backup(){
+    echo "Extracting backups">&2
+    tar xvf harbor.tgz
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:28]:
```suggestion
    tar -xvf harbor.tgz -C ${backup_dir}/
```

Or maybe the backup from S3 always have this folder?

File: scripts/restore/restore-harbor.sh [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+cleanup_local_files(){
+  echo "Cleaning up local files">&2
+  rm harbor.tgz
+  rm ${backup_dir}/registry.back
+  rm ${backup_dir}/postgres.back
+  rm ${backup_dir}/notarysigner.back
+  rm ${backup_dir}/notaryserver.back
+  rmdir ${backup_dir}
```
Comment (Xartos) [inline scripts/restore/restore-harbor.sh:76]:
```suggestion
  rm -r ${backup_dir}
```

**Review by Xartos — CHANGES_REQUESTED** [review]
Overall nice script, but I have some comments

**Comment by Xartos** [comment]
I just read these comments. 
I think we should add the readme from the old instructions. Otherwise it's kind of confusing

**Comment by viktor-f** [comment]
I added the instructions that I'm going to put into the official docs.

@Xartos is it ok that I leave the script as is and instead rework it in a separate issue?

**Comment by viktor-f** [comment]
Added another issue to rework this script: https://github.com/elastisys/compliantkubernetes-apps/issues/386

**Review by Xartos — APPROVED** [review]
@viktor-f Yes, that is fine. That with the addition of the README and I think this is good to go :slightly_smiling_face:

**Pull request merged**